recipes/eleuther_eval.py

```diff
@@ -4,40 +4,279 @@
 
-from torch import nn
-from torchtune import config, generation, training, utils
-from torchtune.data import left_pad_sequence
+from torchtune import config, training, utils
+from torchtune.data import (
+    format_content_with_images,
+    left_pad_sequence,
+    Message,
+    padded_collate_tiled_images_and_mask,
+)
+from torchtune.generation import generate, sample
 from torchtune.modules import TransformerDecoder
+from torchtune.modules.model_fusion import DeepFusionModel
 from torchtune.modules.tokenizers import ModelTokenizer
+from torchtune.modules.transforms import Transform
 from torchtune.recipe_interfaces import EvalRecipeInterface
 
-logger = utils.get_logger("DEBUG")
-
 try:
-    import lm_eval
-    from lm_eval.evaluator import evaluate, get_task_list
-    from lm_eval.models.huggingface import HFLM
-    from lm_eval.tasks import get_task_dict, TaskManager
-    from lm_eval.utils import make_table
-except ImportError:
-    logger.error(
-        "Recipe requires EleutherAI Eval Harness v0.4. Please install with `pip install lm_eval==0.4.*`"
+    lm_eval_version = importlib.metadata.version("lm_eval")
+except importlib.metadata.PackageNotFoundError:
+    lm_eval_version = ""
+
+if not lm_eval_version >= "0.4.2":
+    raise ImportError(
+        "lm_eval version must be >= 0.4.2. Please install lm_eval >= 0.4.2."
     )
-    sys.exit(1)
+
+from lm_eval.evaluator import evaluate, get_task_list
+from lm_eval.models.hf_vlms import HFMultimodalLM
+from lm_eval.models.huggingface import HFLM
+from lm_eval.tasks import get_task_dict, TaskManager
+from lm_eval.utils import make_table
 
 
-class _EvalWrapper(HFLM):
+class _VLMEvalWrapper(HFMultimodalLM):
     """An EvalWrapper for EleutherAI's eval harness based on gpt-fast's
     EvalWrapper: https://github.com/pytorch-labs/gpt-fast/blob/main/eval.py.
 
+    Note:
+        This is ONLY for vision-language models.
+
+    Args:
+        model (DeepFusionModel): The VLM to evaluate.
+        transform (Transform): The transform (tokenizer) to use for preprocessing.
+        device (torch.device): The device to use.
+        max_seq_length (int): The maximum sequence length.
+        batch_size (int): The batch size per GPU.
+        dtype (torch.dtype): dtype for the model caches during generation.
+        enable_kv_cache (bool): Whether to enable KV cache for generation.
+        image_str (str): The string to use for the image token. Default is "<image>", which
+            is the default used by the MMMU dataset.
+        max_images_per_sample (int): The maximum number of images per sample. Defaults to
+            the max number of images in MMMU.
+    """
+
+    def __init__(
+        self,
+        model: DeepFusionModel,
+        transform: Transform,
+        *,
+        device: torch.device,
+        max_seq_length: int = 4096,
+        batch_size: int = 8,
+        dtype: torch.dtype = torch.bfloat16,
+        enable_kv_cache: bool = True,
+        image_str: str = "<image>",
```

Comment:
Not going to make image_st and max_images_per_sample configurable from the config ATM b/c we only support MMMU and these are the features necessary for MMMU.